CMSgov/mint-app
# Discussion Tagging BE

# EASI-3286

## Changes and Description

- Implemented HTML sanitization
   -  tags allowed based on previous work on EASI
   - span allowed with attributes needed for a mention
- Interface / Union type to returned a tagged entity (currently a user account or a possible solution)
- Plan Discussion Created emails take HTML in the body
- Possible Operational Solution Get By key added
   - There is the potential to turn this into a dataloader, but I left this out to avoid scope creep
   
   ### Tag Structure
   `<span data-type="mention" tag-type="USER_ACCOUNT" class="mention" data-id="SKZO" data-label=\"Alexander Stark">@Alexander Stark</span>`
   - `data-type` --> must be "mention"
   - `tag-type` --> `USER_ACCOUNT` or `POSSIBLE_SOLUTION`
   - `class` --> must be "mention"
   - `data-id` --> the key for the data you are tagging, either a username, or a possible operational solution key
   - `data-label` --> Not required by the backend, but sent by the front end library
   - `data-id-db` --> The primary key of the tagged entity in the database. If this is set from the frontend, the backend won't lookup the information or update the tag
   
   After being processed by the backend, the tag is updated to include the `data-id-db` (the pk from the database)
   `<span data-type="mention" tag-type="USER_ACCOUNT" class="mention" data-id="SKZO" data-id-db="f0884065-1b56-40f3-a65b-0c85204e06b7" data-label="Alexander Stark">@Alexander Stark</span>`
   
### High Level Flow
1. Anything with the `scalar` type `TaggedHTMLInput` get's sanitized, and the mentions are parsed using regex.
2. The resolvers get the database primary key for the tagged entity, and update the rawContent of the Tagged html to include `data-id-db`.
3. The updated content is saved to the database in the relevant table (currently only plan discussion and discussion_reply)
4. The HTMLMentions are converted to Tags which are saved to the database.

### Returning tagged entities
1. Using a Union type in GQL, and an interface in GO, a possible operational solution or a user account is conditionally returned based on the tags provided table name, field name, and relevant key.




## How to test this change
1. Create a model Plan
2. Create a Discussion using the Postman Collection
    a. Look at the created email in MailCatcher to see the formatting for the mentions class
3. Create a Discussion reply using the Postman Collection

I'd also like feedback on error handling when creating tags. I suggest creating some bad tags / perhaps with EUAs that can't be found with local lookup. Do you agree with the approach or have suggestions for improvements? 


## PR Author Review Checklist



## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.
